fe/fe-core/src/main/java/com/starrocks/sql/ast/UserVariable.java

```diff
@@ -97,55 +94,86 @@ public String toSql() {
-                        resultExpr = new NullLiteral();
-                    } else {
-                        Type userVariableType = unevaluatedExpression.getType();
-                        //JSON type will be stored as string type
-                        if (userVariableType.isJsonType()) {
-                            userVariableType = Type.VARCHAR;
-                        }
-                        resultExpr = LiteralExpr.create(
-                                StandardCharsets.UTF_8.decode(result.get(0).result).toString(), userVariableType);
-                    }
-                }
-                evaluatedExpression = resultExpr;
-            } catch (TException | AnalysisException e) {
-                throw new SemanticException(e.getMessage());
-            }
+            evaluatedExpression = decodeVariableData(sqlResult.first, unevaluatedExpression.getType());
         }
 
         if (ctx.getState().getStateType() == QueryState.MysqlStateType.ERR) {
             throw new SemanticException(ctx.getState().getErrorMessage());
         }
     }
-    private static List<TVariableData> deserializerVariableData(List<TResultBatch> sqlResult) throws TException {
-        List<TVariableData> statistics = Lists.newArrayList();
-
-        TDeserializer deserializer = new TDeserializer(new TCompactProtocol.Factory());
-        for (TResultBatch resultBatch : sqlResult) {
-            for (ByteBuffer bb : resultBatch.rows) {
-                TVariableData sd = new TVariableData();
-                byte[] bytes = new byte[bb.limit() - bb.position()];
-                bb.get(bytes);
-                deserializer.deserialize(sd, bytes);
-                statistics.add(sd);
+
+    private static Expr decodeVariableData(List<TResultBatch> sqlResult, Type targetType) {
+        Preconditions.checkState(sqlResult.size() == 1,
+                "subquery in user variable must return at most 1 value");
+        Preconditions.checkState(sqlResult.get(0).getRows().size() == 1,
+                "subquery in user variable must return at most 1 value");
+        ByteBuffer byteBuffer = sqlResult.get(0).getRows().get(0);
+        byte[] bytes = new byte[byteBuffer.limit() - byteBuffer.position()];
+        byteBuffer.get(bytes);
+
+        int lengthOffset = getOffset(bytes);
+
+        // process null value
+        if (lengthOffset == -1) {
+            return NullLiteral.create(Type.VARCHAR);
+        }
+
+        String value;
+        try {
+            value = new String(bytes, lengthOffset, bytes.length - lengthOffset, StandardCharsets.UTF_8);
+        } catch (Exception e) {
+            throw new SemanticException("unable decode bytes to string. bytes offset: %s, bytes length: %s",
+                    lengthOffset, bytes.length);
+        }
+
+        //JSON type will be stored as string type
+        if (targetType.isJsonType()) {
+            targetType = Type.VARCHAR;
+        }
+
+        if (targetType.isScalarType()) {
```

Comment:
Json need special process too

Reply:
We won't generate a json type. We just convert the json type to varchar type before this step. This is the old way and I think we don't need change it.